tabular.lua

```diff
@@ -164,6 +164,7 @@ show_as_columns = function(t, bgcolor, seen, column_order, skip_header)
       if type(row) == "table" then
          for k, v in pairs(row) do
             if not column_set or column_set[k] then
+               k = tostring(k)
```

Comment:
Thank you!

This is the only `tostring` that is really necessary. The other two will get the right values as long as this one is in place.

At first I wondered: how did tl not catch this? 

So let's look at `tabular.tl`: 

* In line 148, the type of `t` is declared as `{ {string:any} }` (array of maps of strings to anything)
* so in line 163, the type of `row` should be `{ string:any }`
* in line 164, we're double-checking that `row` is really a `table` at runtime... Suspicious, heh? :thinking: 
* in line 165 when we do `pairs(row)`, if `row` is `{string:any}` then `k` must always be a `string`! why is `tostring` needed?

The answer is in the place where `show_as_columns` is called... in line 269:

* `t as {{string:any}}`! A type cast! The Number One way of hiding mistakes in typed programs. :) This is forcing tl to think and accept that the row keys are strings when they might as well not be.

If I change that to `t as {{any:any}}`, which is really all that the runtime check in line 268 is guaranteeing, I run `tl tabular.tl` again and I get:

```
1 error:
../tabular/tabular.tl:269:37: argument 1: got any, expected string
```

Ok, it's complaining there because the function expects a `{{string:any}}`. Let's change the function signature to be more truthful about what we're really passing it and see what happens.

```
../tabular/tabular.tl:165:17: cannot use operator or for types boolean and <invalid type>
../tabular/tabular.tl:165:32: wrong index type: any, expected string
../tabular/tabular.tl:166:18: wrong index type: any, expected string
../tabular/tabular.tl:167:13: wrong index type: any, expected string
../tabular/tabular.tl:167:26: in assignment: got {}, expected <invalid type>
../tabular/tabular.tl:168:13: cannot index something that is not a record: <invalid type>
../tabular/tabular.tl:168:39: argument 1: got any, expected string
../tabular/tabular.tl:171:11: in assignment: got Output, expected <invalid type>
../tabular/tabular.tl:172:11: cannot index something that is not a record: <invalid type>
../tabular/tabular.tl:172:39: argument 1: got <invalid type>, expected number
```

Whoa! Tons of errors! But they all seem to derive from the fact that k is an `any` and not a string!

Let's add this line as `local k = tostring(k)` (the `local` is there so that a new variable of type `string` replaces the variable of type `any`)

and...

```
Type checked ../tabular/tabular.tl
0 errors detected -- you can use:

   tl run ../tabular/tabular.tl

       to run ../tabular/tabular.tl as a program

   tl gen ../tabular/tabular.tl

       to generate ../tabular/tabular.lua
```

Success!

I'm going to commit a slightly modifed version of this PR with the type changes so that they are more "truthful" (i.e. less reliant on a type casts to compile). Moral of the story: type casts are useful to tell the compiler to "shut up" and get going, but often the compiler is really trying to tell you something important. :)

Reply:
Ahhhh got it now! The type checker looks very promising. I will give a try and start using it in some projects.